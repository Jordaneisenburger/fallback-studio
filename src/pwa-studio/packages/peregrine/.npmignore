.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
