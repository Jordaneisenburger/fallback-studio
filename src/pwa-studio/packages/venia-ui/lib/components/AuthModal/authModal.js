import React from 'react';
import { func, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import CreateAccount from '../CreateAccount';
import ForgotPassword from '../ForgotPassword';
import MyAccount from '../MyAccount';
import SignIn from '../SignIn';
import defaultClasses from './authModal.css';
import { useAuthModal } from '@magento/peregrine/lib/talons/AuthModal/useAuthModal';
import SIGN_OUT_MUTATION from '../../queries/signOut.graphql';

const AuthModal = props => {
    const {
        handleClose,
        handleCreateAccount,
        handleSignOut,
        setUsername,
        showCreateAccount,
        showForgotPassword,
        showMyAccount,
        username
    } = useAuthModal({
        ...props,
        signOutMutation: SIGN_OUT_MUTATION
    });

    let child = null;
    switch (props.view) {
        case 'CREATE_ACCOUNT': {
            child = (
                <CreateAccount
                    initialValues={{ email: username }}
                    onSubmit={handleCreateAccount}
                />
            );
            break;
        }
        case 'FORGOT_PASSWORD': {
            child = (
                <ForgotPassword
                    initialValues={{ email: username }}
                    onClose={handleClose}
                />
            );
            break;
        }
        case 'MY_ACCOUNT': {
            child = <MyAccount onSignOut={handleSignOut} />;
            break;
        }
        case 'SIGN_IN': {
            child = (
                <SignIn
                    setDefaultUsername={setUsername}
                    showCreateAccount={showCreateAccount}
                    showForgotPassword={showForgotPassword}
                    showMyAccount={showMyAccount}
                />
            );
            break;
        }
    }

    const classes = mergeClasses(defaultClasses, props.classes);
    return <div className={classes.root}>{child}</div>;
};

export default AuthModal;

AuthModal.propTypes = {
    classes: shape({
        root: string
    }),
    showCreateAccount: func.isRequired,
    showForgotPassword: func.isRequired,
    showMainMenu: func.isRequired,
    showMyAccount: func.isRequired,
    view: string.isRequired
};
