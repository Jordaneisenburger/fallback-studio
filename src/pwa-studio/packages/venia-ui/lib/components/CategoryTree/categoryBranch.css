.root {
    composes: root from './categoryLeaf.css';
}

.target {
    composes: target from './categoryLeaf.css';
}

.text {
    composes: text from './categoryLeaf.css';
}
