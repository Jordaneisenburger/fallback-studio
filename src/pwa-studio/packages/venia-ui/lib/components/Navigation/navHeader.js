import React, { Fragment } from 'react';
import { bool, func, shape, string } from 'prop-types';
import {
    ArrowLeft as ArrowLeftIcon,
    Menu as MenuIcon,
    X as CloseIcon
} from 'react-feather';

import { mergeClasses } from '../../classify';
import Icon from '../Icon';
import Trigger from '../Trigger';
import defaultClasses from './navHeader.css';
import { useNavigationHeader } from '@magento/peregrine/lib/talons/Navigation/useNavigationHeader';

const titles = {
    CREATE_ACCOUNT: 'Create Account',
    FORGOT_PASSWORD: 'Forgot Password',
    MY_ACCOUNT: 'My Account',
    SIGN_IN: 'Sign In',
    MENU: 'Main Menu'
};

const NavHeader = props => {
    const { isTopLevel, onBack, onClose, view } = props;

    const talonProps = useNavigationHeader({
        isTopLevel,
        onBack,
        onClose,
        view
    });

    const { handleClose, handleBack, isTopLevelMenu } = talonProps;
    const title = titles[view] || titles.MENU;
    const backIcon = isTopLevelMenu ? MenuIcon : ArrowLeftIcon;
    const backButton = !isTopLevelMenu ? (
        <Trigger key="backButton" action={handleBack}>
            <Icon src={backIcon} />
        </Trigger>
    ) : null;

    const classes = mergeClasses(defaultClasses, props.classes);
    return (
        <Fragment>
            {backButton}
            <h2 key="title" className={classes.title}>
                <span>{title}</span>
            </h2>
            <Trigger key="closeButton" action={handleClose}>
                <Icon src={CloseIcon} />
            </Trigger>
        </Fragment>
    );
};

export default NavHeader;

NavHeader.propTypes = {
    classes: shape({
        title: string
    }),
    isTopLevel: bool,
    onBack: func.isRequired,
    onClose: func.isRequired,
    view: string.isRequired
};
