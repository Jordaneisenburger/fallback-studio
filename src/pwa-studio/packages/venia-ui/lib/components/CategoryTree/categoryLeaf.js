import React from 'react';
import { func, shape, string } from 'prop-types';
import { useCategoryLeaf } from '@magento/peregrine/lib/talons/CategoryTree';

import { mergeClasses } from '../../classify';
import { Link, resourceUrl } from '../../drivers';
import defaultClasses from './categoryLeaf.css';

const suffix = '.html';

const Leaf = props => {
    const { category, onNavigate } = props;
    const { name, url_path } = category;
    const classes = mergeClasses(defaultClasses, props.classes);
    const { handleClick } = useCategoryLeaf({ onNavigate });
    const destination = resourceUrl(`/${url_path}${suffix}`);

    return (
        <li className={classes.root}>
            <Link
                className={classes.target}
                to={destination}
                onClick={handleClick}
            >
                <span className={classes.text}>{name}</span>
            </Link>
        </li>
    );
};

export default Leaf;

Leaf.propTypes = {
    category: shape({
        name: string.isRequired,
        url_path: string.isRequired
    }).isRequired,
    classes: shape({
        root: string,
        target: string,
        text: string
    }),
    onNavigate: func.isRequired
};
