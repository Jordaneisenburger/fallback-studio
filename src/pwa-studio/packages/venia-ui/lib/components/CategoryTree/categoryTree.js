import React from 'react';
import { func, number, objectOf, shape, string } from 'prop-types';
import { useCategoryTree } from '@magento/peregrine/lib/talons/CategoryTree';

import { mergeClasses } from '../../classify';
import MENU_QUERY from '../../queries/getNavigationMenu.graphql';
import Branch from './categoryBranch';
import Leaf from './categoryLeaf';
import defaultClasses from './categoryTree.css';

const Tree = props => {
    const {
        categories,
        categoryId,
        onNavigate,
        setCategoryId,
        updateCategories
    } = props;

    const talonProps = useCategoryTree({
        categories,
        categoryId,
        query: MENU_QUERY,
        updateCategories
    });

    const { childCategories } = talonProps;
    const classes = mergeClasses(defaultClasses, props.classes);

    // for each child category, render a direct link if it has no children
    // otherwise render a branch
    const branches = Array.from(childCategories, childCategory => {
        const [id, { category, isLeaf }] = childCategory;

        return isLeaf ? (
            <Leaf key={id} category={category} onNavigate={onNavigate} />
        ) : (
            <Branch
                key={id}
                category={category}
                setCategoryId={setCategoryId}
            />
        );
    });

    return (
        <div className={classes.root}>
            <ul className={classes.tree}>{branches}</ul>
        </div>
    );
};

export default Tree;

Tree.propTypes = {
    categories: objectOf(
        shape({
            id: number.isRequired,
            name: string
        })
    ),
    categoryId: number.isRequired,
    classes: shape({
        root: string,
        tree: string
    }),
    onNavigate: func.isRequired,
    setCategoryId: func.isRequired,
    updateCategories: func.isRequired
};
