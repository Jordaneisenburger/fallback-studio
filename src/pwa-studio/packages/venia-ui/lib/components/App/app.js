import React, { useCallback, useEffect } from 'react';
import { array, func, shape, string } from 'prop-types';

import { useToasts } from '@magento/peregrine';
import { useApp } from '@magento/peregrine/lib/talons/App/useApp';

import { HeadProvider, Title } from '../Head';
import Main from '../Main';
import Mask from '../Mask';
import MiniCart from '../MiniCart';
import Navigation from '../Navigation';
import Routes from '../Routes';
import { registerMessageHandler } from '../../util/swUtils';
import { HTML_UPDATE_AVAILABLE } from '../../constants/swMessageTypes';
import ToastContainer from '../ToastContainer';
import Icon from '../Icon';

import {
    AlertCircle as AlertCircleIcon,
    CloudOff as CloudOffIcon,
    Wifi as WifiIcon,
    RefreshCcw as RefreshIcon
} from 'react-feather';

const OnlineIcon = <Icon src={WifiIcon} attrs={{ width: 18 }} />;
const OfflineIcon = <Icon src={CloudOffIcon} attrs={{ width: 18 }} />;
const ErrorIcon = <Icon src={AlertCircleIcon} attrs={{ width: 18 }} />;
const UpdateIcon = <Icon src={RefreshIcon} attrs={{ width: 18 }} />;

const ERROR_MESSAGE = 'Sorry! An unexpected error occurred.';

const App = props => {
    const { markErrorHandled, renderError, unhandledErrors } = props;

    const [, { addToast }] = useToasts();

    const handleIsOffline = useCallback(() => {
        addToast({
            type: 'error',
            icon: OfflineIcon,
            message: 'You are offline. Some features may be unavailable.',
            timeout: 3000
        });
    }, [addToast]);

    const handleIsOnline = useCallback(() => {
        addToast({
            type: 'info',
            icon: OnlineIcon,
            message: 'You are online.',
            timeout: 3000
        });
    }, [addToast]);

    const handleHTMLUpdate = useCallback(
        resetHTMLUpdateAvaiableFlag => {
            addToast({
                type: 'warning',
                icon: UpdateIcon,
                message: 'Update available. Please refresh.',
                actionText: 'Refresh',
                timeout: 0,
                onAction: () => {
                    location.reload();
                },
                onDismiss: removeToast => {
                    resetHTMLUpdateAvaiableFlag();
                    removeToast();
                }
            });
        },
        [addToast]
    );

    const handleError = useCallback(
        (error, id, loc, handleDismissError) => {
            const errorToastProps = {
                icon: ErrorIcon,
                message: `${ERROR_MESSAGE}\nDebug: ${id} ${loc}`,
                onDismiss: remove => {
                    handleDismissError();
                    remove();
                },
                timeout: 15000,
                type: 'error'
            };

            addToast(errorToastProps);
        },
        [addToast]
    );

    const talonProps = useApp({
        handleError,
        handleIsOffline,
        handleIsOnline,
        handleHTMLUpdate,
        markErrorHandled,
        renderError,
        unhandledErrors
    });

    const {
        hasOverlay,
        handleCloseDrawer,
        setHTMLUpdateAvailable
    } = talonProps;

    useEffect(() => {
        const unregisterHandler = registerMessageHandler(
            HTML_UPDATE_AVAILABLE,
            () => {
                setHTMLUpdateAvailable(true);
            }
        );
        return unregisterHandler;
    }, [setHTMLUpdateAvailable]);

    if (renderError) {
        return (
            <HeadProvider>
                <Title>{`Home Page - ${STORE_NAME}`}</Title>
                <Main isMasked={true} />
                <Mask isActive={true} />
                <ToastContainer />
            </HeadProvider>
        );
    }

    return (
        <HeadProvider>
            <Title>{`Home Page - ${STORE_NAME}`}</Title>
            <Main isMasked={hasOverlay}>
                <Routes />
            </Main>
            <Mask isActive={hasOverlay} dismiss={handleCloseDrawer} />
            <Navigation />
            <MiniCart />
            <ToastContainer />
        </HeadProvider>
    );
};

App.propTypes = {
    markErrorHandled: func.isRequired,
    renderError: shape({
        stack: string
    }),
    unhandledErrors: array
};

export default App;
