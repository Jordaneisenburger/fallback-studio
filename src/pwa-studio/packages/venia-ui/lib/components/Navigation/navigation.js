import React, { Suspense } from 'react';
import { shape, string } from 'prop-types';

import { useNavigation } from '@magento/peregrine/lib/talons/Navigation/useNavigation';

import { mergeClasses } from '../../classify';
import AuthBar from '../AuthBar';
import CategoryTree from '../CategoryTree';
import LoadingIndicator from '../LoadingIndicator';
import NavHeader from './navHeader';
import defaultClasses from './navigation.css';
import GET_CUSTOMER_QUERY from '../../queries/getCustomer.graphql';

const AuthModal = React.lazy(() => import('../AuthModal'));

const Navigation = props => {
    const {
        catalogActions,
        categories,
        categoryId,
        handleBack,
        handleClose,
        hasModal,
        isOpen,
        isTopLevel,
        setCategoryId,
        showCreateAccount,
        showForgotPassword,
        showMainMenu,
        showMyAccount,
        showSignIn,
        view
    } = useNavigation({ customerQuery: GET_CUSTOMER_QUERY });

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClassName = isOpen ? classes.root_open : classes.root;
    const modalClassName = hasModal ? classes.modal_open : classes.modal;
    const bodyClassName = hasModal ? classes.body_masked : classes.body;
    const rootHeaderClassName =
        isTopLevel && view === 'MENU' ? classes.isRoot : classes.header;

    // Lazy load the auth modal because it may not be needed.
    const authModal = hasModal ? (
        <Suspense fallback={<LoadingIndicator />}>
            <AuthModal
                closeDrawer={handleClose}
                showCreateAccount={showCreateAccount}
                showForgotPassword={showForgotPassword}
                showMainMenu={showMainMenu}
                showMyAccount={showMyAccount}
                showSignIn={showSignIn}
                view={view}
            />
        </Suspense>
    ) : null;

    return (
        <aside className={rootClassName}>
            <header className={rootHeaderClassName}>
                <NavHeader
                    isTopLevel={isTopLevel}
                    onBack={handleBack}
                    onClose={handleClose}
                    view={view}
                />
            </header>
            <div className={bodyClassName}>
                <CategoryTree
                    categoryId={categoryId}
                    categories={categories}
                    onNavigate={handleClose}
                    setCategoryId={setCategoryId}
                    updateCategories={catalogActions.updateCategories}
                />
            </div>
            <div className={classes.footer}>
                <AuthBar
                    disabled={hasModal}
                    showMyAccount={showMyAccount}
                    showSignIn={showSignIn}
                />
            </div>
            <div className={modalClassName}>{authModal}</div>
        </aside>
    );
};

export default Navigation;

Navigation.propTypes = {
    classes: shape({
        body: string,
        form_closed: string,
        form_open: string,
        footer: string,
        header: string,
        root: string,
        root_open: string,
        signIn_closed: string,
        signIn_open: string,
        isRoot: string
    })
};
