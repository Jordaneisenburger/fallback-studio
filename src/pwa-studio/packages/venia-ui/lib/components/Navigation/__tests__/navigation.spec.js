import React from 'react';
import { act } from 'react-test-renderer';
import { createTestInstance } from '@magento/peregrine';

import { useAppContext } from '@magento/peregrine/lib/context/app';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import AuthModal from '../../AuthModal';
import CategoryTree from '../../CategoryTree';
import NavHeader from '../navHeader';
import Navigation from '../navigation';

jest.mock('../../../classify');
jest.mock('../../AuthBar', () => () => <i />);
jest.mock('../../AuthModal', () => () => <i />);
jest.mock('../../CategoryTree', () => () => <i />);
jest.mock('../navHeader', () => () => <i />);

jest.mock('@magento/peregrine/lib/context/app', () => {
    const closeDrawer = jest.fn();
    const useAppContext = jest.fn(() => [
        { drawer: 'nav' },
        {
            actions: {},
            closeDrawer
        }
    ]);

    return { useAppContext };
});

jest.mock('@magento/peregrine/lib/context/catalog', () => {
    const updateCategories = jest.fn();
    const useCatalogContext = jest.fn(() => [
        {
            categories: {
                1: { parentId: 0 },
                2: { parentId: 1 }
            },
            rootCategoryId: 1
        },
        {
            actions: { updateCategories }
        }
    ]);

    return { useCatalogContext };
});

jest.mock('@magento/peregrine/lib/context/user', () => {
    const getUserDetails = jest.fn();
    const useUserContext = jest.fn(() => [{}, { getUserDetails }]);

    return { useUserContext };
});

test('renders correctly when open', () => {
    const instance = createTestInstance(<Navigation />);

    expect(instance.toJSON()).toMatchSnapshot();
    expect(instance.root.findByProps({ className: 'root_open' })).toBeTruthy();
});

test('renders correctly when closed', () => {
    useAppContext.mockImplementationOnce(() => [
        { drawer: null },
        { closeDrawer: jest.fn() }
    ]);

    const instance = createTestInstance(<Navigation />);

    expect(instance.toJSON()).toMatchSnapshot();
    expect(instance.root.findByProps({ className: 'root' })).toBeTruthy();
});

test('getUserDetails() is called on mount', () => {
    const { getUserDetails } = useUserContext()[1];

    createTestInstance(<Navigation />);

    expect(getUserDetails).toHaveBeenCalledTimes(1);
});

test('showCreateAccount updates the view', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showCreateAccount } = root.findByType(AuthModal).props;

    act(() => {
        showCreateAccount();
    });

    const header = root.findByType(NavHeader);

    expect(header.props.view).toBe('CREATE_ACCOUNT');
});

test('showForgotPassword updates the view', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showForgotPassword } = root.findByType(AuthModal).props;

    act(() => {
        showForgotPassword();
    });

    const header = root.findByType(NavHeader);

    expect(header.props.view).toBe('FORGOT_PASSWORD');
});

test('showMainMenu updates the view', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showMainMenu } = root.findByType(AuthModal).props;

    act(() => {
        showMainMenu();
    });

    const header = root.findByType(NavHeader);

    expect(header.props.view).toBe('MENU');
});

test('showMyAccount updates the view', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showMyAccount } = root.findByType(AuthModal).props;

    act(() => {
        showMyAccount();
    });

    const header = root.findByType(NavHeader);

    expect(header.props.view).toBe('MY_ACCOUNT');
});

test('showSignIn updates the view', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showSignIn } = root.findByType(AuthModal).props;

    act(() => {
        showSignIn();
    });

    const header = root.findByType(NavHeader);

    expect(header.props.view).toBe('SIGN_IN');
});

test('back button closes the drawer', () => {
    const { closeDrawer } = useAppContext()[1];
    const { root } = createTestInstance(<Navigation />);
    const { onBack: handleBack } = root.findByType(NavHeader).props;

    act(() => {
        handleBack();
    });

    expect(closeDrawer).toHaveBeenCalledTimes(1);
});

test('back button returns from MyAccount to menu', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showMyAccount } = root.findByType(AuthModal).props;

    act(() => {
        showMyAccount();
    });

    act(() => {
        const { onBack: handleBack } = root.findByType(NavHeader).props;

        handleBack();
    });

    expect(root.findByType(AuthModal).props.view).toBe('MENU');
});

test('back button returns from SignIn to menu', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showSignIn } = root.findByType(AuthModal).props;

    act(() => {
        showSignIn();
    });

    act(() => {
        const { onBack: handleBack } = root.findByType(NavHeader).props;

        handleBack();
    });

    expect(root.findByType(AuthModal).props.view).toBe('MENU');
});

test('back button returns from CreateAccount to SignIn', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showCreateAccount } = root.findByType(AuthModal).props;

    act(() => {
        showCreateAccount();
    });

    act(() => {
        const { onBack: handleBack } = root.findByType(NavHeader).props;

        handleBack();
    });

    expect(root.findByType(AuthModal).props.view).toBe('SIGN_IN');
});

test('back button returns from ForgotPassword to SignIn', () => {
    const { root } = createTestInstance(<Navigation />);
    const { showForgotPassword } = root.findByType(AuthModal).props;

    act(() => {
        showForgotPassword();
    });

    act(() => {
        const { onBack: handleBack } = root.findByType(NavHeader).props;

        handleBack();
    });

    expect(root.findByType(AuthModal).props.view).toBe('SIGN_IN');
});

test('back button returns to the parent category', () => {
    const { root } = createTestInstance(<Navigation />);
    const { setCategoryId } = root.findByType(CategoryTree).props;

    act(() => {
        setCategoryId(2);
    });

    expect(root.findByType(CategoryTree).props.categoryId).toBe(2);

    act(() => {
        const { onBack: handleBack } = root.findByType(NavHeader).props;

        handleBack();
    });

    expect(root.findByType(CategoryTree).props.categoryId).toBe(1);
});
