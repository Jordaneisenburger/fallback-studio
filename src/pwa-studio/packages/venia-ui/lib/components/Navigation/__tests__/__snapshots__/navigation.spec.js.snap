// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when closed 1`] = `
<aside
  className="root"
>
  <header
    className="isRoot"
  >
    <i />
  </header>
  <div
    className="body"
  >
    <i />
  </div>
  <div
    className="footer"
  >
    <i />
  </div>
  <div
    className="modal"
  >
    <i />
  </div>
</aside>
`;

exports[`renders correctly when open 1`] = `
<aside
  className="root_open"
>
  <header
    className="isRoot"
  >
    <i />
  </header>
  <div
    className="body"
  >
    <i />
  </div>
  <div
    className="footer"
  >
    <i />
  </div>
  <div
    className="modal"
  >
    <i />
  </div>
</aside>
`;
